pkg/storageops/azure/azure.go

```diff
@@ -0,0 +1,808 @@
+	if err != nil {
+		return nil, err
+	}
+
+	var disks []interface{}
+	for _, id := range diskNames {
+		if d, ok := allDisks[*id]; ok {
+			disks = append(disks, d)
+		} else {
+			return nil, storageops.NewStorageError(
+				storageops.ErrVolNotFound,
+				fmt.Sprintf("disk %s not found", *id),
+				a.instance,
+			)
+		}
+	}
+
+	return disks, nil
+}
+
+func (a *azureOps) DeviceMappings() (map[string]string, error) {
+	vm, err := a.describeInstance()
+	if err != nil {
+		return nil, err
+	}
+
+	devMap := make(map[string]string)
+	for _, d := range *vm.StorageProfile.DataDisks {
+		devPath, err := lunToBlockDevPath(*d.Lun)
+		if err != nil {
+			return nil, storageops.NewStorageError(
+				storageops.ErrInvalidDevicePath,
+				fmt.Sprintf("unable to find block dev path for lun%v: %v", *d.Lun, err),
+				a.instance,
+			)
+		}
+		devMap[devPath] = *d.Name
+	}
+
+	return devMap, nil
+}
+
+func (a *azureOps) Enumerate(
+	diskNames []*string,
+	labels map[string]string,
+	setIdentifier string,
+) (map[string][]interface{}, error) {
+	allDisks, err := a.getDisks(labels)
+	if err != nil {
+		return nil, err
+	}
+
+	sets := make(map[string][]interface{})
+	for _, disk := range allDisks {
+		if len(setIdentifier) == 0 {
+			storageops.AddElementToMap(sets, disk, storageops.SetIdentifierNone)
+		} else {
+			found := false
+			for key, value := range disk.Tags {
+				if key == setIdentifier && value != nil {
+					storageops.AddElementToMap(sets, disk, *value)
+					found = true
+					break
+				}
+			}
+
+			if !found {
+				storageops.AddElementToMap(sets, disk, storageops.SetIdentifierNone)
+			}
+		}
+	}
+
+	return sets, nil
+}
+
+func (a *azureOps) DevicePath(diskName string) (string, error) {
+	disk, err := a.disksClient.Get(
+		context.Background(),
+		a.resourceGroupName,
+		diskName,
+	)
+	if err != nil {
+		return "", err
+	}
+
+	if disk.ManagedBy == nil || len(*disk.ManagedBy) == 0 {
+		return "", storageops.NewStorageError(
+			storageops.ErrVolDetached,
+			fmt.Sprintf("Disk: %s is detached", diskName),
+			a.instance,
+		)
+	}
+
+	vm, err := a.describeInstance()
+	if err != nil {
+		return "", err
+	}
+
+	if vm.StorageProfile == nil || vm.StorageProfile.DataDisks == nil {
+		return "", fmt.Errorf("instance does not have any disks attached")
+	}
+
+	for _, d := range *vm.StorageProfile.DataDisks {
+		if *d.Name == diskName {
+			devPath, err := lunToBlockDevPathWithRetry(*d.Lun)
+			if err == nil {
+				return devPath, nil
+			}
+			return "", storageops.NewStorageError(
+				storageops.ErrInvalidDevicePath,
+				fmt.Sprintf("unable to find block dev path for lun%v: %v", *d.Lun, err),
+				a.instance,
+			)
+		}
+	}
+
+	return "", storageops.NewStorageError(
+		storageops.ErrVolAttachedOnRemoteNode,
+		fmt.Sprintf("disk %s is not attached on: %s (Attached on: %v)",
+			diskName, a.instance, disk.ManagedBy),
+		a.instance,
+	)
+}
+
+func (a *azureOps) Snapshot(diskName string, readonly bool) (interface{}, error) {
```

Comment:
Looks like read-write snapshots aren't supported for Azure, so setting `readonly` to false should return an error.